tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -850,5 +850,543 @@ int main(int argc, char **argv)
+                    .count = s2n_array_len(test_cipher_suite_list),
+                    .suites = test_cipher_suite_list,
+                    .allow_chacha20_boosting = false,
+                };
+
+                uint8_t wire[] = {
+                    /* Client signalled chacha20 boosting. Negotiated as this is the only negotiable option. */
+                    TLS_CHACHA20_POLY1305_SHA256,
+                };
+                uint8_t count = sizeof(wire) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                /* Verify server can still negotiate chacha20 if it's the only option even when chacha20 boosting is off */
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, wire, count));
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_tls13_chacha20_poly1305_sha256);
+            }
+
+            /*
+             * sslv2 server correctly negotiates ciphersuite when chacha20 boosting is enabled.
+             */
+            {
+                DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(connection);
+
+                connection->security_policy_override = &security_policy;
+                connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+                EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(connection, S2N_SSLv2));
+                EXPECT_SUCCESS(s2n_connection_set_config(connection, server_config));
+
+                static struct s2n_cipher_suite *test_cipher_suite_list[] = {
+                    /* First 'valid' ciphersuite that is saved as a higher_vers_match and is negotiated */
+                    &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha,                  
+                    /* Not valid for connection (does not meet minimum version requirement and higher_vers_match already found) */
+                    &s2n_dhe_rsa_with_chacha20_poly1305_sha256,
+                    /* Not valid for connection (does not meet minimum version requirement and higher_vers_match already found) */
+                    &s2n_dhe_rsa_with_aes_256_gcm_sha384
+                };
+
+                /* cppcheck-suppress redundantAssignment */
+                cipher_preferences = (struct s2n_cipher_preferences) {
+                    .count = s2n_array_len(test_cipher_suite_list),
+                    .suites = test_cipher_suite_list,
+                    .allow_chacha20_boosting = true,
+                };
+
+                uint8_t wire[] = {
+                    /* Client signalled chacha20 boosting. Not negotiated as it's not offered by the server. */
+                    0x00,
+                    TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+                    /* Saved by server as higher_vers_match; negotiated as a 'fall-back' ciphersuite */
+                    0x00,
+                    TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
```

Comment:
Don't we actually negotiate s2n_ecdhe_ecdsa_with_aes_128_cbc_sha?

Reply:
Oops, good catch. Fixed comment